fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedFieldInfosFormat.java

```diff
@@ -20,28 +20,67 @@
 
 package com.apple.foundationdb.record.lucene.codec;
 
+import com.apple.foundationdb.record.lucene.directory.FDBDirectory;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import org.apache.commons.lang3.ArrayUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.apache.lucene.codecs.FieldInfosFormat;
 import org.apache.lucene.index.FieldInfos;
+import org.apache.lucene.index.IndexFileNames;
 import org.apache.lucene.index.SegmentInfo;
 import org.apache.lucene.store.Directory;
 import org.apache.lucene.store.IOContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 import java.io.IOException;
+import java.util.BitSet;
+import java.util.List;
 
 /**
  * This class attempts to reduce the number of .fnm entries when numerous segments are present in FDB.
  *
  */
 public class LuceneOptimizedFieldInfosFormat extends FieldInfosFormat {
+    public static final int DEFAULT_MAXIMUM_FIELD_INFO_CACHE_SIZE = 128;
+    private static final Logger LOGGER = LoggerFactory.getLogger(LuceneOptimizedFieldInfosFormat.class);
 
     private final FieldInfosFormat fieldInfosFormat;
 
+    private final Cache<Pair<Directory, BitSet>, FieldInfos> fieldInfosCache;
```

Comment:
What is the lifecycle of `LuceneOptimizedFieldInfosFormat`? Is it created once per transaction? Once per instance? Can it support multiple FDBDirectories (with multiple `subspace`s)?

Reply:
Singleton:     public LuceneOptimizedCodec(Lucene87Codec.Mode mode) {
        super("RL");
        baseCodec = new Lucene87Codec(mode);
        compoundFormat = new LuceneOptimizedCompoundFormat();
        segmentInfoFormat = new LuceneOptimizedSegmentInfoFormat();
        pointsFormat = new LuceneOptimizedPointsFormat(baseCodec.pointsFormat());
        defaultPostingsFormat = new LuceneOptimizedPostingsFormat(new Lucene84PostingsFormat());
        defaultDocValuesFormat = new LuceneOptimizedDocValuesFormat(new Lucene80DocValuesFormat());
        storedFieldsFormat = new LuceneOptimizedStoredFieldsFormat(baseCodec.storedFieldsFormat());
        liveDocsFormat = new LuceneOptimizedLiveDocsFormat(baseCodec.liveDocsFormat());
        fieldInfosFormat = new LuceneOptimizedFieldInfosFormat(new LuceneOptimized60FieldInfosFormat());
    }

Reply:
I think I fixed this with my prior commit and moving the cache off of the LuceneOptimizedFieldInfosFormat?